client/lib/happychat/connection.js

```diff
@@ -5,53 +5,63 @@
  */
 
 import IO from 'socket.io-client';
-import { EventEmitter } from 'events';
-import config from 'config';
 import { v4 as uuid } from 'uuid';
+import { isString } from 'lodash';
 
 /**
  * Internal dependencies
  */
-import { HAPPYCHAT_MESSAGE_TYPES } from 'state/happychat/constants';
+import {
+	HAPPYCHAT_MESSAGE_TYPES
+} from 'state/happychat/constants';
+import {
+	receiveChatEvent,
+	requestChatTranscript,
+	setConnected,
+	setConnecting,
+	setDisconnected,
+	setHappychatAvailable,
+	setHappychatChatStatus,
+	setReconnecting,
+} from 'state/happychat/actions';
 
-/*
- * Happychat client connection for Socket.IO
- */
-import debugFactory from 'debug';
-
-const debug = debugFactory( 'calypso:happychat:connection' );
-
-class Connection extends EventEmitter {
-	open( signer_user_id, jwt, locale, groups ) {
-		if ( ! this.openSocket ) {
-			this.openSocket = new Promise( resolve => {
-				const url = config( 'happychat_url' );
-				const socket = new IO( url );
-				socket
-					.once( 'connect', () => debug( 'connected' ) )
-					.on( 'init', () => {
-						this.emit( 'connected' );
-						resolve( socket );
-					} )
-					.on( 'token', handler => {
-						handler( { signer_user_id, jwt, locale, groups } );
-					} )
-					.on( 'unauthorized', () => {
-						socket.close();
-						debug( 'not authorized' );
-					} )
-					.on( 'disconnect', reason => this.emit( 'disconnect', reason ) )
-					.on( 'reconnecting', () => this.emit( 'reconnecting' ) )
-					// Received a chat message
-					.on( 'message', message => this.emit( 'message', message ) )
-					// Received chat status new/assigning/assigned/missed/pending/abandoned
-					.on( 'status', status => this.emit( 'status', status ) )
-					// If happychat is currently accepting chats
-					.on( 'accept', accept => this.emit( 'accept', accept ) );
-			} );
-		} else {
-			debug( 'socket already initiaized' );
+const debug = require( 'debug' )( 'calypso:happychat:connection' );
+
+const buildConnection = ( socket ) => isString( socket )
+	? new IO( socket ) // If socket is an URL, connect to server.
+	: socket; // If socket is not an url, use it directly. Useful for testing.
+
+class Connection {
+
+	init( url, dispatch, { signer_user_id, jwt, locale, groups, geo_location } ) {
+		if ( this.openSocket ) {
+			debug( 'socket is already connected' );
+			return this.openSocket;
 		}
+
+		dispatch( setConnecting() );
+
+		const socket = buildConnection( url );
+		this.openSocket = new Promise( resolve => {
+			// TODO: reject this promise
```

Comment:
You could probably just do

```js
socket
    .once( 'connect', () => debug( 'connected' ) )
	.on( 'token', handler => handler( { signer_user_id, jwt, locale, groups } ) )
	.on( 'init', () => {
        ...
        // prevent future errors from triggering the reject on this promise
        socket. removeListener( 'error', reject );
        resolve( socket );
    } )
    .on( 'error', reject )

Reply:
Omar, could you elaborate when the server sends an `error` event? 

I still need to review the socket auth process, so I may be wrong here: I was thinking about rejecting the Promise upon the `unauthenticated` event because I thought that was the final event the server sent when the authentication wasn't successful. Is that right?

Reply:
I'd rather introduce this behavior change in a separate PR, so it's easy to revert in case it's necessary.